community/primitive-collections/src/test/java/org/neo4j/concurrent/WorkSyncTest.java

```diff
@@ -257,4 +305,192 @@ public void mustNotApplyWorkInParallelUnderStress() throws Exception
+        };
+
+        List<Future<Void>> futureList = new ArrayList<>();
+        for ( int i = 0; i < workers; i++ )
+        {
+            futureList.add( executor.submit( work ) );
+        }
+        startLatch.await();
+        start.set( true );
+        endLatch.await();
+
+        for ( Future<Void> future : futureList )
+        {
+            future.get(); // check for any exceptions
+        }
+
+        assertThat( count.sum(), lessThan( (long) (workers * iterations) ) );
+        assertThat( sum.sum(), is( (long) (incrementValue * workers * iterations) ) );
+    }
+
+    @Test
+    public void mustApplyWorkAsync() throws Exception
+    {
+        AsyncApply a = sync.applyAsync( new AddWork( 10 ) );
+        a.await();
+        assertThat( sum.sum(), is( 10L ) );
+
+        AsyncApply b = sync.applyAsync( new AddWork( 20 ) );
+        AsyncApply c = sync.applyAsync( new AddWork( 30 ) );
+        b.await();
+        c.await();
+        assertThat( sum.sum(), is( 60L ) );
+    }
+
+    @Test
+    public void mustCombineWorkAsync() throws Exception
+    {
+        makeWorkStuckAtSemaphore( 1 );
+
+        AsyncApply a = sync.applyAsync( new AddWork( 1 ) );
+        AsyncApply b = sync.applyAsync( new AddWork( 1 ) );
+        AsyncApply c = sync.applyAsync( new AddWork( 1 ) );
+        semaphore.release( 2 );
+        a.await();
+        b.await();
+        c.await();
+        assertThat( sum.sum(), is( 4L ) );
+        assertThat( count.sum(), is( 2L ) );
+    }
+
+    @Test
+    public void mustApplyWorkAsyncEvenWhenInterrupted() throws Exception
+    {
+        Thread.currentThread().interrupt();
+
+        sync.applyAsync( new AddWork( 10 ) ).await();
+
+        assertThat( sum.sum(), is( 10L ) );
+        assertTrue( Thread.interrupted() );
+    }
+
+    @Test
+    public void asyncWorkThatThrowsMustRememberException() throws Exception
+    {
+        RuntimeException boo = new RuntimeException( "boo" );
+        AsyncApply asyncApply = sync.applyAsync( new AddWork( 10 )
+        {
+            @Override
+            public void apply( Adder adder )
+            {
+                super.apply( adder );
+                throw boo;
+            }
+        } );
+
+        try
+        {
+            asyncApply.await();
+            fail( "Should have thrown" );
+        }
+        catch ( ExecutionException e )
+        {
+            assertThat( e.getCause(), sameInstance( boo ) );
+        }
+
+        assertThat( sum.sum(), is( 10L ) );
+        assertThat( count.sum(), is( 1L ) );
+
+        try
+        {
+            asyncApply.await();
+            fail( "Should have thrown" );
+        }
+        catch ( ExecutionException e )
+        {
+            assertThat( e.getCause(), sameInstance( boo ) );
+        }
+
+        assertThat( sum.sum(), is( 10L ) );
+        assertThat( count.sum(), is( 1L ) );
+    }
+
+    @Test
+    public void asyncWorkDoneInAnotherThreadThatThrowsMustRememberException() throws Exception // TODO!!!
```

Comment:
Could you clarify how there's another thread in this test throwing an exception which that main test thread (calling `await`) gets to see?